Objects/object.c

```diff
@@ -2091,3 +2140,73 @@ _Py_Dealloc(PyObject *op)
 #ifdef __cplusplus
 }
 #endif
+
+/* TODO(twouters): refactor into a separate file */
+
+/* Remote refcount management */
+
+#define REFCOUNTLINKS_PER_PAGE 32768
+#define LAST_REFCOUNTLINK (REFCOUNTLINKS_PER_PAGE - 1)
+
+typedef union refcountlink {
+  Py_ssize_t refcount;
+  union refcountlink *link;
+} refcountlink;
+
+refcountlink *next_free = NULL;
+refcountlink *first_page = NULL;
+refcountlink *last_page = NULL;
+
+static futex_t refcountlinks_lock = FUTEX_STATIC_INIT("refcountlinks lock");
+
+static void
+alloc_refcount_links(void)
+{
+    ptrdiff_t diff;
+    int i;
+    refcountlink *new_page;
+
+    new_page = PyMem_MALLOC(REFCOUNTLINKS_PER_PAGE * sizeof(refcountlink));
+    if (new_page == NULL) {
+        Py_FatalError("Unable to allocate memory for refcounts");
+    }
+    if (first_page == NULL)
+        first_page = new_page;
+    if (last_page != NULL)
+        last_page[LAST_REFCOUNTLINK].link = new_page;
+    new_page[0].link = last_page;
+    last_page = new_page;
+    diff = &new_page[1] - &new_page[0];
+    for (i = 1; i < REFCOUNTLINKS_PER_PAGE - 1; i++)
+        new_page[i].link = &new_page[i] + diff;
+    new_page[LAST_REFCOUNTLINK].link = NULL;
+    next_free = &new_page[1];
+}
+
+void
+_PyRefcount_New(PyObject *ob)
+{
+    refcountlink *link;
+    futex_lock(&refcountlinks_lock);
+    if (next_free == NULL)
+        alloc_refcount_links();
+    if (next_free == NULL)
+        abort();
+    link = next_free;
+    next_free = link->link;
+    ob->ob_refcnt_ptr = &(link->refcount);
+    Py_SET_REFCNT(ob, 1);
+    futex_unlock(&refcountlinks_lock);
```

Comment:
It could, but then you can have refcounts of value 0 in the used refcounts. Current code doesn't care, but I was thinking of adding a debug mode that did.